// @forestadmin/datasource-toolkit/dist/src/decorators/empty/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const collection_decorator_1 = __importDefault(require("../collection-decorator"));
const branch_1 = __importDefault(require("../../interfaces/query/condition-tree/nodes/branch"));
const leaf_1 = __importDefault(require("../../interfaces/query/condition-tree/nodes/leaf"));
/**
 * Avoid performing useless database requests:
 * Using segments + scopes + filters + jointure emulation often yields requests which have
 * mutually exclusive conditions or empty "In" leafs.
 */
class EmptyCollectionDecorator extends collection_decorator_1.default {
    async list(caller, filter, projection) {
        if (!this.returnsEmptySet(filter.conditionTree)) {
            return super.list(caller, filter, projection);
        }
        return [];
    }
    async update(caller, filter, patch) {
        if (!this.returnsEmptySet(filter.conditionTree)) {
            return super.update(caller, filter, patch);
        }
    }
    async delete(caller, filter) {
        if (!this.returnsEmptySet(filter.conditionTree)) {
            return super.delete(caller, filter);
        }
    }
    async aggregate(caller, filter, aggregation, limit) {
        if (!this.returnsEmptySet(filter.conditionTree)) {
            return super.aggregate(caller, filter, aggregation, limit);
        }
        return [];
    }
    returnsEmptySet(tree) {
        if (tree instanceof leaf_1.default) {
            return this.leafReturnsEmptySet(tree);
        }
        if (tree instanceof branch_1.default && tree.aggregator === 'Or') {
            return this.orReturnsEmptySet(tree.conditions);
        }
        if (tree instanceof branch_1.default && tree.aggregator === 'And') {
            return this.andReturnsEmptySet(tree.conditions);
        }
        return false;
    }
    leafReturnsEmptySet(leaf) {
        // Empty 'in` always return zero records.
        return leaf.operator === 'In' && leaf.value.length === 0;
    }
    orReturnsEmptySet(conditions) {
        // Or return no records when
        // - they have no conditions
        // - they have only conditions which return zero records.
        return conditions.length === 0 || conditions.every(c => this.returnsEmptySet(c));
    }
    andReturnsEmptySet(conditions) {
        // There is a leaf which returns zero records
        if (conditions.some(c => this.returnsEmptySet(c))) {
            return true;
        }
        // Scans for mutually exclusive conditions
        // (this a naive implementation, it will miss many occurences)
        const valuesByField = {};
        const leafs = conditions.filter(condition => condition instanceof leaf_1.default);
        for (const { field, operator, value } of leafs) {
            if (!valuesByField[field] && operator === 'Equal') {
                valuesByField[field] = [value];
            }
            else if (!valuesByField[field] && operator === 'In') {
                valuesByField[field] = value;
            }
            else if (valuesByField[field] && operator === 'Equal') {
                valuesByField[field] = valuesByField[field].includes(value) ? [value] : [];
            }
            else if (valuesByField[field] && operator === 'In') {
                valuesByField[field] = valuesByField[field].filter(v => value.includes(v));
            }
        }
        return Object.values(valuesByField).some(v => v.length === 0);
    }
}
exports.default = EmptyCollectionDecorator;
