// @forestadmin/datasource-toolkit/dist/src/decorators/collection-decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class CollectionDecorator {
    constructor(childCollection, dataSource) {
        this.childCollection = childCollection;
        this.dataSource = dataSource;
    }
    get schema() {
        const subSchema = this.childCollection.schema;
        if (!this.lastSchema || this.lastSubSchema !== subSchema) {
            this.lastSchema = this.refineSchema(subSchema);
            this.lastSubSchema = subSchema;
        }
        return this.lastSchema;
    }
    get name() {
        return this.childCollection.name;
    }
    async execute(caller, name, data, filter) {
        const refinedFilter = await this.refineFilter(caller, filter);
        return this.childCollection.execute(caller, name, data, refinedFilter);
    }
    async getForm(caller, name, data, filter) {
        const refinedFilter = await this.refineFilter(caller, filter);
        return this.childCollection.getForm(caller, name, data, refinedFilter);
    }
    async create(caller, data) {
        return this.childCollection.create(caller, data);
    }
    async list(caller, filter, projection) {
        const refinedFilter = await this.refineFilter(caller, filter);
        return this.childCollection.list(caller, refinedFilter, projection);
    }
    async update(caller, filter, patch) {
        const refinedFilter = await this.refineFilter(caller, filter);
        return this.childCollection.update(caller, refinedFilter, patch);
    }
    async delete(caller, filter) {
        const refinedFilter = await this.refineFilter(caller, filter);
        return this.childCollection.delete(caller, refinedFilter);
    }
    async aggregate(caller, filter, aggregation, limit) {
        const refinedFilter = await this.refineFilter(caller, filter);
        return this.childCollection.aggregate(caller, refinedFilter, aggregation, limit);
    }
    markSchemaAsDirty() {
        this.lastSchema = null;
    }
    async refineFilter(caller, filter) {
        return filter;
    }
    refineSchema(subSchema) {
        return subSchema;
    }
}
exports.default = CollectionDecorator;
