// @forestadmin/datasource-toolkit/dist/src/decorators/datasource-decorator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_datasource_1 = __importDefault(require("../base-datasource"));
class DataSourceDecorator extends base_datasource_1.default {
    constructor(childDataSource, CollectionDecoratorCtor) {
        super();
        this.addCollectionToChildDataSource = childDataSource.addCollection.bind(childDataSource);
        Reflect.defineProperty(childDataSource, 'addCollection', {
            value: this.addCollectionObserver.bind(this),
        });
        this.childDataSource = childDataSource;
        this.CollectionDecoratorCtor = CollectionDecoratorCtor;
        this.childDataSource.collections.forEach(collection => this.addCollection(new this.CollectionDecoratorCtor(collection, this)));
    }
    get schema() {
        return this.childDataSource.schema;
    }
    renderChart(caller, name) {
        return this.childDataSource.renderChart(caller, name);
    }
    addCollectionObserver(collection) {
        this.addCollectionToChildDataSource(collection);
        this.addCollection(new this.CollectionDecoratorCtor(collection, this));
    }
}
exports.default = DataSourceDecorator;
