// @forestadmin/datasource-toolkit/dist/src/decorators/search/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const uuid_1 = require("uuid");
const collection_context_1 = __importDefault(require("../../context/collection-context"));
const collection_decorator_1 = __importDefault(require("../collection-decorator"));
const factory_1 = __importDefault(require("../../interfaces/query/condition-tree/factory"));
const leaf_1 = __importDefault(require("../../interfaces/query/condition-tree/nodes/leaf"));
class SearchCollectionDecorator extends collection_decorator_1.default {
    constructor() {
        super(...arguments);
        this.replacer = null;
    }
    replaceSearch(replacer) {
        this.replacer = replacer;
    }
    refineSchema(subSchema) {
        return { ...subSchema, searchable: true };
    }
    async refineFilter(caller, filter) {
        // Search string is not significant
        if (!filter?.search || SearchCollectionDecorator.checkEmptyString(filter.search)) {
            return filter.override({ search: null });
        }
        // Implement search ourselves
        if (this.replacer || !this.childCollection.schema.searchable) {
            const ctx = new collection_context_1.default(this, caller);
            let tree = this.defaultReplacer(filter.search, filter.searchExtended);
            if (this.replacer) {
                const plainTree = await this.replacer(filter.search, filter.searchExtended, ctx);
                tree = factory_1.default.fromPlainObject(plainTree);
            }
            // Note that if no fields are searchable with the provided searchString, the conditions
            // array might be empty, which will create a condition returning zero records
            // (this is the desired behavior).
            return filter.override({
                conditionTree: factory_1.default.intersect(filter.conditionTree, tree),
                search: null,
            });
        }
        // Let subcollection deal with the search
        return filter;
    }
    defaultReplacer(search, extended) {
        const searchableFields = SearchCollectionDecorator.getSearchFields(this.childCollection.schema, this.childCollection.dataSource, extended);
        const conditions = searchableFields
            .map(([field, schema]) => SearchCollectionDecorator.buildCondition(field, schema, search))
            .filter(Boolean);
        return factory_1.default.union(...conditions);
    }
    static buildCondition(field, schema, searchString) {
        const { columnType, enumValues } = schema;
        const isNumber = Number(searchString).toString() === searchString;
        const isUuid = (0, uuid_1.validate)(searchString);
        if (columnType === 'Number' && isNumber) {
            return new leaf_1.default(field, 'Equal', Number(searchString));
        }
        if (columnType === 'Enum') {
            const searchValue = SearchCollectionDecorator.lenientFind(enumValues, searchString);
            if (searchValue)
                return new leaf_1.default(field, 'Equal', searchValue);
        }
        if (columnType === 'String') {
            return new leaf_1.default(field, 'IContains', searchString);
        }
        if (columnType === 'Uuid' && isUuid) {
            return new leaf_1.default(field, 'Equal', searchString);
        }
        return null;
    }
    static lenientFind(enumValues, searchString) {
        return (enumValues?.find(v => v === searchString.trim()) ??
            enumValues?.find(v => v.toLocaleLowerCase() === searchString.toLocaleLowerCase().trim()));
    }
    static getSearchFields(schema, dataSource, searchExtended) {
        const fields = Object.entries(schema.fields);
        if (searchExtended)
            SearchCollectionDecorator.getDeepFields(dataSource, fields);
        return fields.filter(([, fieldSchema]) => SearchCollectionDecorator.isSearchable(fieldSchema));
    }
    static getDeepFields(dataSource, fields) {
        fields.forEach(([name, field]) => {
            if (field.type === 'ManyToOne' || field.type === 'OneToOne') {
                const related = dataSource.getCollection(field.foreignCollection);
                fields.push(...Object.entries(related.schema.fields).map(([subName, columnSchema]) => [`${name}:${subName}`, columnSchema]));
            }
        });
    }
    static checkEmptyString(searchString) {
        return searchString.trim().length === 0;
    }
    static isSearchable(schema) {
        if (schema.type === 'Column') {
            const { columnType, filterOperators } = schema;
            if (columnType === 'Enum' || columnType === 'Number' || columnType === 'Uuid') {
                return filterOperators?.has('Equal');
            }
            if (columnType === 'String') {
                return filterOperators?.has('IContains');
            }
        }
        return false;
    }
}
exports.default = SearchCollectionDecorator;
