// @forestadmin/datasource-toolkit/dist/src/decorators/actions/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_1 = __importDefault(require("./context/base"));
const single_1 = __importDefault(require("./context/single"));
const collection_decorator_1 = __importDefault(require("../collection-decorator"));
const result_builder_1 = __importDefault(require("./result-builder"));
class ActionCollectionDecorator extends collection_decorator_1.default {
    constructor() {
        super(...arguments);
        this.actions = {};
    }
    addAction(name, action) {
        this.actions[name] = action;
        this.markSchemaAsDirty();
    }
    async execute(caller, name, data, filter) {
        const action = this.actions[name];
        if (!action)
            return this.childCollection.execute(caller, name, data, filter);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const context = this.getContext(caller, action, data, filter);
        const resultBuilder = new result_builder_1.default();
        const result = await action.execute(context, resultBuilder);
        return (result || {
            type: 'Success',
            invalidated: new Set(),
            format: 'text',
            message: 'Success',
        });
    }
    async getForm(caller, name, data, filter) {
        const action = this.actions[name];
        if (!action)
            return this.childCollection.getForm(caller, name, data, filter);
        if (!action.form)
            return [];
        const formValues = data ? { ...data } : {};
        const used = new Set();
        const context = this.getContext(caller, action, formValues, filter, used);
        // Convert DynamicField to ActionField in successive steps.
        let dynamicFields;
        dynamicFields = action.form.map(c => ({ ...c }));
        dynamicFields = await this.dropDefaults(context, dynamicFields, !data, formValues);
        dynamicFields = await this.dropIfs(context, dynamicFields);
        const fields = await this.dropDeferred(context, dynamicFields);
        for (const field of fields) {
            // customer did not define a handler to rewrite the previous value => reuse current one.
            if (field.value === undefined)
                field.value = formValues[field.label];
            // fields that were accessed through the context.formValues.X getter should be watched.
            field.watchChanges = used.has(field.label);
        }
        return fields;
    }
    refineSchema(subSchema) {
        const newSchema = { ...subSchema, actions: { ...subSchema.actions } };
        for (const [name, { form, scope, generateFile }] of Object.entries(this.actions)) {
            // An action form can be send in the schema to avoid calling the load handler
            // as long as there is nothing dynamic in it.
            const isDynamic = form?.some(field => Object.values(field).some(value => typeof value === 'function'));
            newSchema.actions[name] = { scope, generateFile: !!generateFile, staticForm: !isDynamic };
        }
        return newSchema;
    }
    getContext(caller, action, formValues, filter, used) {
        return new {
            Global: base_1.default,
            Bulk: base_1.default,
            Single: single_1.default,
        }[action.scope](this, caller, formValues, filter, used);
    }
    async dropDefaults(context, fields, isFirstCall, data) {
        if (isFirstCall) {
            const defaults = await Promise.all(fields.map(field => this.evaluate(context, field.defaultValue)));
            fields.forEach((field, index) => {
                data[field.label] = defaults[index];
            });
        }
        fields.forEach(field => delete field.defaultValue);
        return fields;
    }
    async dropIfs(context, fields) {
        // Remove fields which have falsy if
        const ifValues = await Promise.all(fields.map(field => !field.if || this.evaluate(context, field.if)));
        const newFields = fields.filter((_, index) => ifValues[index]);
        newFields.forEach(field => delete field.if);
        return newFields;
    }
    async dropDeferred(context, fields) {
        const newFields = fields.map(async (field) => {
            const keys = Object.keys(field);
            const values = await Promise.all(Object.values(field).map(value => this.evaluate(context, value)));
            return keys.reduce((memo, key, index) => ({ ...memo, [key]: values[index] }), {});
        });
        return Promise.all(newFields);
    }
    async evaluate(context, value) {
        return typeof value === 'function' ? value(context) : value;
    }
}
exports.default = ActionCollectionDecorator;
