// @forestadmin/datasource-toolkit/dist/src/base-collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class BaseCollection {
    constructor(name, datasource) {
        this.dataSource = null;
        this.name = null;
        this.schema = null;
        this.dataSource = datasource;
        this.name = name;
        this.schema = {
            actions: {},
            countable: false,
            fields: {},
            searchable: false,
            segments: [],
        };
    }
    addAction(name, schema) {
        const action = this.schema.actions[name];
        if (action !== undefined)
            throw new Error(`Action "${name}" already defined in collection`);
        this.schema.actions[name] = schema;
    }
    addField(name, schema) {
        const fieldSchema = this.schema.fields[name];
        if (fieldSchema !== undefined)
            throw new Error(`Field "${name}" already defined in collection`);
        this.schema.fields[name] = schema;
    }
    addFields(fields) {
        Object.entries(fields).forEach(([fieldName, fieldSchema]) => this.addField(fieldName, fieldSchema));
    }
    addSegments(segments) {
        this.schema.segments.push(...segments);
    }
    enableCount() {
        this.schema.countable = true;
    }
    enableSearch() {
        this.schema.searchable = true;
    }
    async execute(caller, name) {
        throw new Error(this.schema.actions[name]
            ? `Action ${name} is not implemented.`
            : `No action named ${name} was added`);
    }
    async getForm() {
        return [];
    }
}
exports.default = BaseCollection;
