// @forestadmin/datasource-toolkit/dist/src/composite-datasource.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_datasource_1 = __importDefault(require("./base-datasource"));
class CompositeDatasource extends base_datasource_1.default {
    constructor() {
        super();
        this.datasourceChartMapping = {};
    }
    addDataSource(dataSource) {
        this.addCollections(dataSource);
        this.addCharts(dataSource);
        return this;
    }
    get schema() {
        return { charts: Object.keys(this.datasourceChartMapping) };
    }
    renderChart(caller, name) {
        if (!this.datasourceChartMapping[name]) {
            throw new Error(`Chart '${name}' is not defined in the dataSource.`);
        }
        return this.datasourceChartMapping[name].renderChart(caller, name);
    }
    addCollections(dataSource) {
        for (const collection of dataSource.collections) {
            this.addCollection(collection);
        }
    }
    addCharts(dataSource) {
        for (const chartName of dataSource.schema.charts) {
            if (this.datasourceChartMapping[chartName]) {
                throw new Error(`Chart '${chartName}' is already defined in datasource.`);
            }
            this.datasourceChartMapping[chartName] = dataSource;
        }
    }
}
exports.default = CompositeDatasource;
