// @forestadmin/datasource-toolkit/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaUtils = exports.RecordUtils = exports.CollectionUtils = exports.SortValidator = exports.RecordValidator = exports.ProjectionValidator = exports.FieldValidator = exports.ConditionTreeValidator = exports.SortFactory = exports.Sort = exports.ProjectionFactory = exports.Projection = exports.PaginatedFilter = exports.Page = exports.FilterFactory = exports.Filter = exports.ConditionTreeLeaf = exports.ConditionTreeFactory = exports.ConditionTreeBranch = exports.ConditionTree = exports.Aggregation = exports.HookCollectionDecorator = exports.SortEmulateCollectionDecorator = exports.SegmentCollectionDecorator = exports.SchemaCollectionDecorator = exports.WriteCollectionDecorator = exports.SearchCollectionDecorator = exports.RenameFieldCollectionDecorator = exports.PublicationCollectionDecorator = exports.OperatorsReplaceCollectionDecorator = exports.OperatorsEmulateCollectionDecorator = exports.RelationCollectionDecorator = exports.EmptyCollectionDecorator = exports.ComputedCollectionDecorator = exports.ActionCollectionDecorator = exports.RenameCollectionDataSourceDecorator = exports.ChartDataSourceDecorator = exports.DataSourceDecorator = exports.CollectionCustomizationContext = exports.CompositeDatasource = exports.BaseDataSource = exports.BaseCollection = void 0;
// Misc
__exportStar(require("./errors"), exports);
__exportStar(require("./factory"), exports);
// Base Collection & DataSource
var base_collection_1 = require("./base-collection");
Object.defineProperty(exports, "BaseCollection", { enumerable: true, get: function () { return __importDefault(base_collection_1).default; } });
var base_datasource_1 = require("./base-datasource");
Object.defineProperty(exports, "BaseDataSource", { enumerable: true, get: function () { return __importDefault(base_datasource_1).default; } });
var composite_datasource_1 = require("./composite-datasource");
Object.defineProperty(exports, "CompositeDatasource", { enumerable: true, get: function () { return __importDefault(composite_datasource_1).default; } });
// Context
var collection_context_1 = require("./context/collection-context");
Object.defineProperty(exports, "CollectionCustomizationContext", { enumerable: true, get: function () { return __importDefault(collection_context_1).default; } });
// Decorators (datasource)
var datasource_decorator_1 = require("./decorators/datasource-decorator");
Object.defineProperty(exports, "DataSourceDecorator", { enumerable: true, get: function () { return __importDefault(datasource_decorator_1).default; } });
var datasource_1 = require("./decorators/chart/datasource");
Object.defineProperty(exports, "ChartDataSourceDecorator", { enumerable: true, get: function () { return __importDefault(datasource_1).default; } });
var datasource_2 = require("./decorators/rename-collection/datasource");
Object.defineProperty(exports, "RenameCollectionDataSourceDecorator", { enumerable: true, get: function () { return __importDefault(datasource_2).default; } });
// Decorators (collections)
var collection_1 = require("./decorators/actions/collection");
Object.defineProperty(exports, "ActionCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_1).default; } });
var collection_2 = require("./decorators/computed/collection");
Object.defineProperty(exports, "ComputedCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_2).default; } });
var collection_3 = require("./decorators/empty/collection");
Object.defineProperty(exports, "EmptyCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_3).default; } });
var collection_4 = require("./decorators/relation/collection");
Object.defineProperty(exports, "RelationCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_4).default; } });
var collection_5 = require("./decorators/operators-emulate/collection");
Object.defineProperty(exports, "OperatorsEmulateCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_5).default; } });
var collection_6 = require("./decorators/operators-replace/collection");
Object.defineProperty(exports, "OperatorsReplaceCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_6).default; } });
var collection_7 = require("./decorators/publication/collection");
Object.defineProperty(exports, "PublicationCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_7).default; } });
var collection_8 = require("./decorators/rename-field/collection");
Object.defineProperty(exports, "RenameFieldCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_8).default; } });
var collection_9 = require("./decorators/search/collection");
Object.defineProperty(exports, "SearchCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_9).default; } });
var collection_10 = require("./decorators/write/collection");
Object.defineProperty(exports, "WriteCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_10).default; } });
var collection_11 = require("./decorators/schema/collection");
Object.defineProperty(exports, "SchemaCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_11).default; } });
var collection_12 = require("./decorators/segment/collection");
Object.defineProperty(exports, "SegmentCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_12).default; } });
var collection_13 = require("./decorators/sort-emulate/collection");
Object.defineProperty(exports, "SortEmulateCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_13).default; } });
var collection_14 = require("./decorators/hook/collection");
Object.defineProperty(exports, "HookCollectionDecorator", { enumerable: true, get: function () { return __importDefault(collection_14).default; } });
// Query Interface
var aggregation_1 = require("./interfaces/query/aggregation");
Object.defineProperty(exports, "Aggregation", { enumerable: true, get: function () { return __importDefault(aggregation_1).default; } });
var base_1 = require("./interfaces/query/condition-tree/nodes/base");
Object.defineProperty(exports, "ConditionTree", { enumerable: true, get: function () { return __importDefault(base_1).default; } });
var branch_1 = require("./interfaces/query/condition-tree/nodes/branch");
Object.defineProperty(exports, "ConditionTreeBranch", { enumerable: true, get: function () { return __importDefault(branch_1).default; } });
var factory_1 = require("./interfaces/query/condition-tree/factory");
Object.defineProperty(exports, "ConditionTreeFactory", { enumerable: true, get: function () { return __importDefault(factory_1).default; } });
var leaf_1 = require("./interfaces/query/condition-tree/nodes/leaf");
Object.defineProperty(exports, "ConditionTreeLeaf", { enumerable: true, get: function () { return __importDefault(leaf_1).default; } });
var unpaginated_1 = require("./interfaces/query/filter/unpaginated");
Object.defineProperty(exports, "Filter", { enumerable: true, get: function () { return __importDefault(unpaginated_1).default; } });
var factory_2 = require("./interfaces/query/filter/factory");
Object.defineProperty(exports, "FilterFactory", { enumerable: true, get: function () { return __importDefault(factory_2).default; } });
var page_1 = require("./interfaces/query/page");
Object.defineProperty(exports, "Page", { enumerable: true, get: function () { return __importDefault(page_1).default; } });
var paginated_1 = require("./interfaces/query/filter/paginated");
Object.defineProperty(exports, "PaginatedFilter", { enumerable: true, get: function () { return __importDefault(paginated_1).default; } });
var projection_1 = require("./interfaces/query/projection");
Object.defineProperty(exports, "Projection", { enumerable: true, get: function () { return __importDefault(projection_1).default; } });
var factory_3 = require("./interfaces/query/projection/factory");
Object.defineProperty(exports, "ProjectionFactory", { enumerable: true, get: function () { return __importDefault(factory_3).default; } });
var sort_1 = require("./interfaces/query/sort");
Object.defineProperty(exports, "Sort", { enumerable: true, get: function () { return __importDefault(sort_1).default; } });
var factory_4 = require("./interfaces/query/sort/factory");
Object.defineProperty(exports, "SortFactory", { enumerable: true, get: function () { return __importDefault(factory_4).default; } });
__exportStar(require("./interfaces/action"), exports);
__exportStar(require("./interfaces/caller"), exports);
__exportStar(require("./interfaces/collection"), exports);
__exportStar(require("./interfaces/query/aggregation"), exports);
__exportStar(require("./interfaces/query/condition-tree/nodes/base"), exports);
__exportStar(require("./interfaces/query/condition-tree/nodes/branch"), exports);
__exportStar(require("./interfaces/query/condition-tree/nodes/leaf"), exports);
__exportStar(require("./interfaces/query/filter/paginated"), exports);
__exportStar(require("./interfaces/query/filter/unpaginated"), exports);
__exportStar(require("./interfaces/query/sort"), exports);
__exportStar(require("./interfaces/record"), exports);
__exportStar(require("./interfaces/schema"), exports);
__exportStar(require("./interfaces/templates"), exports);
// Validation
var condition_tree_1 = require("./validation/condition-tree");
Object.defineProperty(exports, "ConditionTreeValidator", { enumerable: true, get: function () { return __importDefault(condition_tree_1).default; } });
var field_1 = require("./validation/field");
Object.defineProperty(exports, "FieldValidator", { enumerable: true, get: function () { return __importDefault(field_1).default; } });
var projection_2 = require("./validation/projection");
Object.defineProperty(exports, "ProjectionValidator", { enumerable: true, get: function () { return __importDefault(projection_2).default; } });
var record_1 = require("./validation/record");
Object.defineProperty(exports, "RecordValidator", { enumerable: true, get: function () { return __importDefault(record_1).default; } });
var sort_2 = require("./validation/sort");
Object.defineProperty(exports, "SortValidator", { enumerable: true, get: function () { return __importDefault(sort_2).default; } });
// Utils
var collection_15 = require("./utils/collection");
Object.defineProperty(exports, "CollectionUtils", { enumerable: true, get: function () { return __importDefault(collection_15).default; } });
var record_2 = require("./utils/record");
Object.defineProperty(exports, "RecordUtils", { enumerable: true, get: function () { return __importDefault(record_2).default; } });
var schema_1 = require("./utils/schema");
Object.defineProperty(exports, "SchemaUtils", { enumerable: true, get: function () { return __importDefault(schema_1).default; } });
