// @forestadmin/datasource-mongoose/dist/utils/pipeline/reparent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mongoose_1 = require("mongoose");
const schema_1 = __importDefault(require("../../mongoose/schema"));
/**
 * Generate pipeline to query submodels.
 *
 * The operations make rotations in the documents so that the root is changed to the submodel
 * without loosing the parent (which may be needed later on).
 */
class ReparentGenerator {
    static reparent(model, prefix) {
        if (!prefix?.length)
            return [];
        // Take only the part of the records we are interested in
        const pipeline = [];
        let schema = schema_1.default.fromModel(model).fields;
        for (const part of prefix.split('.')) {
            pipeline.push(...(schema[part]['[]']
                ? this.reparentArray(part, schema[part]['[]'] instanceof mongoose_1.SchemaType)
                : this.reparentObject(part, schema[part] instanceof mongoose_1.SchemaType)));
            schema = schema[part];
            if (schema['[]'])
                schema = schema['[]'];
        }
        // If we end up on a field, create a virtual record
        return pipeline;
    }
    static reparentArray(prefix, inDoc) {
        return [
            { $unwind: { path: `$${prefix}`, includeArrayIndex: 'index' } },
            {
                $replaceRoot: {
                    newRoot: {
                        $mergeObjects: [
                            inDoc ? { content: `$${prefix}` } : `$${prefix}`,
                            {
                                _id: { $concat: [{ $toString: '$_id' }, `.${prefix}.`, { $toString: '$index' }] },
                                _pid: '$_id',
                                parent: '$$ROOT',
                            },
                        ],
                    },
                },
            },
        ];
    }
    static reparentObject(prefix, inDoc) {
        return [
            {
                $replaceRoot: {
                    newRoot: {
                        $mergeObjects: [
                            inDoc ? { content: `$${prefix}` } : `$${prefix}`,
                            {
                                _id: { $concat: [{ $toString: '$_id' }, `.${prefix}`] },
                                _pid: '$_id',
                                parent: '$$ROOT',
                            },
                        ],
                    },
                },
            },
        ];
    }
}
exports.default = ReparentGenerator;
