// @forestadmin/datasource-mongoose/dist/mongoose/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable no-underscore-dangle */
/* eslint-disable @typescript-eslint/no-explicit-any */
const mongoose_1 = require("mongoose");
const helpers_1 = require("../utils/helpers");
class MongooseSchema {
    constructor(models, fields, isArray, isLeaf) {
        this.models = models;
        this.fields = fields;
        this.isArray = isArray;
        this.isLeaf = isLeaf;
    }
    static fromModel(model) {
        return new MongooseSchema(model.db.models, this.buildFields(model.schema), false, false);
    }
    get schemaType() {
        if (this.fields.content instanceof mongoose_1.SchemaType) {
            return this.fields.content;
        }
        throw new Error(`Schema is not a leaf.`);
    }
    getSubSchema(path, withParent = false) {
        // Terminating condition
        if (path === null)
            return this;
        // General case: go down the tree
        const dotIndex = path.indexOf('.');
        const prefix = dotIndex === -1 ? path : path.substring(0, dotIndex);
        const suffix = dotIndex === -1 ? null : path.substring(dotIndex + 1);
        let isArray = false;
        let isLeaf = false;
        let child = this.fields[prefix];
        // Traverse relations
        if (prefix.endsWith('__manyToOne')) {
            const foreignKeyName = prefix.substring(0, prefix.length - '__manyToOne'.length);
            const relationName = this.fields[foreignKeyName].options.ref;
            child = MongooseSchema.fromModel(this.models[relationName]).fields;
        }
        // Traverse arrays
        if (child['[]']) {
            child = child['[]'];
            isArray = true;
        }
        // We ended up on a field => box it.
        if (child instanceof mongoose_1.SchemaType) {
            child = { content: child };
            isLeaf = true;
        }
        // If we want parent to be included, re-add the rootSchema
        if (withParent && !prefix.endsWith('__manyToOne')) {
            const id = new mongoose_1.Schema.Types.String(prefix, {}, 'String');
            child = { ...child, _id: id, _pid: this.fields._id, parent: this.fields };
        }
        return new MongooseSchema(this.models, child, isArray, isLeaf).getSubSchema(suffix, withParent);
    }
    /**
     * Build a tree of SchemaType from a mongoose schema.
     * This removes most complexity from using prefixes, nested schemas and array types
     */
    static buildFields(schema, level = 0) {
        const paths = {};
        for (const [name, field] of Object.entries(schema.paths)) {
            // Exclude mixedFieldPattern $* and privateFieldPattern __
            if (!name.startsWith('$*') && !name.includes('__') && (name !== '_id' || level === 0)) {
                // Flatten nested schemas and arrays
                if (field.constructor.name === 'SubdocumentPath') {
                    const subPaths = this.buildFields(field.schema, level + 1);
                    for (const [subName, subField] of Object.entries(subPaths))
                        (0, helpers_1.recursiveSet)(paths, `${name}.${subName}`, subField);
                }
                else if (field.constructor.name === 'DocumentArrayPath') {
                    const subPaths = this.buildFields(field.schema, level + 1);
                    for (const [subName, subField] of Object.entries(subPaths))
                        (0, helpers_1.recursiveSet)(paths, `${name}.[].${subName}`, subField);
                }
                else if (field.constructor.name === 'SchemaArray') {
                    (0, helpers_1.recursiveSet)(paths, `${name}.[]`, field.caster);
                }
                else {
                    (0, helpers_1.recursiveSet)(paths, name, field);
                }
            }
        }
        return paths;
    }
}
exports.default = MongooseSchema;
