// @forestadmin/datasource-toolkit/dist/src/validation/condition-tree.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const rules_1 = require("./rules");
const errors_1 = require("../errors");
const collection_1 = __importDefault(require("../utils/collection"));
const branch_1 = __importDefault(require("../interfaces/query/condition-tree/nodes/branch"));
const leaf_1 = __importDefault(require("../interfaces/query/condition-tree/nodes/leaf"));
const field_1 = __importDefault(require("./field"));
const type_getter_1 = __importDefault(require("./type-getter"));
class ConditionTreeValidator {
    static validate(conditionTree, collection) {
        if (conditionTree instanceof branch_1.default) {
            ConditionTreeValidator.validateBranch(conditionTree, collection);
        }
        else if (conditionTree instanceof leaf_1.default) {
            ConditionTreeValidator.validateLeaf(conditionTree, collection);
        }
        else {
            throw new errors_1.ValidationError('Unexpected condition tree type');
        }
    }
    static validateBranch(branch, collection) {
        if (!['And', 'Or'].includes(branch.aggregator)) {
            throw new errors_1.ValidationError(`The given aggregator '${branch.aggregator}' ` +
                `is not supported. The supported values are: ['Or', 'And']`);
        }
        if (!Array.isArray(branch.conditions)) {
            throw new errors_1.ValidationError(`The given conditions '${branch.conditions}' were expected to be an array`);
        }
        for (const condition of branch.conditions) {
            ConditionTreeValidator.validate(condition, collection);
        }
    }
    static validateLeaf(leaf, collection) {
        const fieldSchema = collection_1.default.getFieldSchema(collection, leaf.field);
        ConditionTreeValidator.throwIfOperatorNotAllowedWithColumn(leaf, fieldSchema);
        ConditionTreeValidator.throwIfValueNotAllowedWithOperator(leaf, fieldSchema);
        ConditionTreeValidator.throwIfOperatorNotAllowedWithColumnType(leaf, fieldSchema);
        ConditionTreeValidator.throwIfValueNotAllowedWithColumnType(leaf, fieldSchema);
    }
    static throwIfOperatorNotAllowedWithColumn(conditionTree, columnSchema) {
        const operators = columnSchema.filterOperators;
        if (!operators?.has(conditionTree.operator)) {
            throw new errors_1.ValidationError(`The given operator '${conditionTree.operator}' ` +
                `is not supported by the column: '${conditionTree.field}'.\n${operators?.size
                    ? `The allowed types are: [${[...operators]}]`
                    : 'the column is not filterable'}`);
        }
    }
    static throwIfValueNotAllowedWithOperator(conditionTree, columnSchema) {
        const { value } = conditionTree;
        const valueType = type_getter_1.default.get(value, columnSchema.columnType);
        const allowedTypes = rules_1.MAP_ALLOWED_TYPES_FOR_OPERATOR[conditionTree.operator];
        if (!allowedTypes.includes(valueType)) {
            throw new errors_1.ValidationError(`The given value attribute '${JSON.stringify(value)} (type: ${valueType})' has an unexpected value ` +
                `for the given operator '${conditionTree.operator}'.\n` +
                `${allowedTypes.length === 0
                    ? 'The value attribute must be empty.'
                    : `The allowed types of the field value are: [${allowedTypes}].`}`);
        }
    }
    static throwIfOperatorNotAllowedWithColumnType(conditionTree, columnSchema) {
        const allowedOperators = rules_1.MAP_ALLOWED_OPERATORS_FOR_COLUMN_TYPE[columnSchema.columnType];
        if (!allowedOperators.includes(conditionTree.operator)) {
            throw new errors_1.ValidationError(`The given operator '${conditionTree.operator}' ` +
                `is not allowed with the columnType schema: '${columnSchema.columnType}'.\n` +
                `The allowed types are: [${allowedOperators}]`);
        }
    }
    static throwIfValueNotAllowedWithColumnType(conditionTree, columnSchema) {
        const { value, field } = conditionTree;
        const { columnType } = columnSchema;
        const allowedTypes = rules_1.MAP_ALLOWED_TYPES_FOR_COLUMN_TYPE[columnType];
        field_1.default.validateValue(field, columnSchema, value, allowedTypes);
    }
}
exports.default = ConditionTreeValidator;
