// @forestadmin/datasource-toolkit/dist/src/validation/type-getter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const luxon_1 = require("luxon");
const uuid_1 = require("uuid");
const types_1 = require("./types");
class TypeGetter {
    static get(value, typeContext) {
        if (Array.isArray(value))
            return TypeGetter.getArrayType(value, typeContext);
        if (typeof value === 'string')
            return TypeGetter.getTypeFromString(value, typeContext);
        if (typeof value === 'number' && !Number.isNaN(Number(value)))
            return 'Number';
        if (value instanceof Date && luxon_1.DateTime.fromJSDate(value).isValid)
            return 'Date';
        if (typeof value === 'boolean')
            return 'Boolean';
        if (typeof value === 'object' && typeContext === 'Json')
            return 'Json';
        return types_1.ValidationPrimaryTypes.Null;
    }
    static getArrayType(value, typeContext) {
        if (value.length === 0)
            return types_1.ValidationTypesArray.Empty;
        if (TypeGetter.isArrayOf('Number', value, typeContext))
            return types_1.ValidationTypesArray.Number;
        if (TypeGetter.isArrayOf('Uuid', value, typeContext))
            return types_1.ValidationTypesArray.Uuid;
        if (TypeGetter.isArrayOf('Boolean', value, typeContext))
            return types_1.ValidationTypesArray.Boolean;
        if (TypeGetter.isArrayOf('String', value, typeContext))
            return types_1.ValidationTypesArray.String;
        if (TypeGetter.isArrayOf('Enum', value, typeContext))
            return types_1.ValidationTypesArray.Enum;
        return types_1.ValidationPrimaryTypes.Null;
    }
    static getDateType(value) {
        const dateTime = luxon_1.DateTime.fromISO(value);
        if (dateTime.toISODate() === value)
            return 'Dateonly';
        if (dateTime.toISOTime().match(value))
            return 'Timeonly';
        return 'Date';
    }
    static getTypeFromString(value, typeContext) {
        if (['Enum', 'String'].includes(typeContext))
            return typeContext;
        if ((0, uuid_1.validate)(value))
            return 'Uuid';
        if (TypeGetter.isValidDate(value))
            return TypeGetter.getDateType(value);
        if (TypeGetter.isJson(value))
            return 'Json';
        if (TypeGetter.isPoint(value, typeContext))
            return 'Point';
        return 'String';
    }
    static isValidDate(value) {
        return luxon_1.DateTime.fromISO(value).isValid;
    }
    static isPoint(value, typeContext) {
        const potentialPoint = value.split(',');
        return (potentialPoint.length === 2 &&
            typeContext === 'Point' &&
            TypeGetter.get(potentialPoint.map(Number), 'Number') === types_1.ValidationTypesArray.Number);
    }
    static isJson(value) {
        try {
            return !!JSON.parse(value);
        }
        catch {
            return false;
        }
    }
    static isArrayOf(type, values, typeContext) {
        return values.every(item => TypeGetter.get(item, typeContext) === type);
    }
}
exports.default = TypeGetter;
