// @forestadmin/datasource-toolkit/dist/src/context/relaxed-wrappers/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const aggregation_1 = __importDefault(require("../../interfaces/query/aggregation"));
const factory_1 = __importDefault(require("../../interfaces/query/condition-tree/factory"));
const unpaginated_1 = __importDefault(require("../../interfaces/query/filter/unpaginated"));
const page_1 = __importDefault(require("../../interfaces/query/page"));
const paginated_1 = __importDefault(require("../../interfaces/query/filter/paginated"));
const projection_1 = __importDefault(require("../../interfaces/query/projection"));
const sort_1 = __importDefault(require("../../interfaces/query/sort"));
/** Collection wrapper which accepts plain objects in all methods */
class RelaxedCollection {
    constructor(collection, caller) {
        this.collection = collection;
        this.caller = caller;
    }
    /**
     * Execute a given action
     * @param name the name of the action
     * @param formValues the values of the form, if the action rely on an action form
     * @param filter the filter used to represent the selected records to use the action on
     * @example
     * .execute(
     *    'Refund',
     *    { reason: 'Article is broken' },
     *    {
     *      conditionTree: {
     *        field: 'id',
     *        operator: 'Equal',
     *        value: 1
     *      }
     *    }
     * );
     */
    execute(name, formValues, filter) {
        const filterInstance = this.buildFilter(filter);
        return this.collection.execute(this.caller, name, formValues, filterInstance);
    }
    getForm(name, formValues, filter) {
        const filterInstance = this.buildFilter(filter);
        return this.collection.getForm(this.caller, name, formValues, filterInstance);
    }
    /**
     * Create a list of records
     * @param data An array of records to create
     * @example
     * .create([
     *    { amountInEur: 150, description: 'Buy dvd' },
     *    { amountInEur: -100, description: 'Refund' },
     * ]);
     */
    create(data) {
        return this.collection.create(this.caller, data);
    }
    /**
     * List multiple records
     * @param filter the filter used to select the records to list
     * @param projection an array of fields name representing the data to select
     * @example
     * .list({
     *    conditionTree: {
     *      aggregator: 'And',
     *      conditions: [{
     *        field: 'amountInEur',
     *        operator: 'GreaterThan',
     *        value: 1000
     *      }, {
     *        field: 'description',
     *        operator: 'Contains',
     *        value: 'Refund',
     *      }],
     *      page: { limit: 10, skip: 0 },
     *      sort: [{ field: 'id', ascending: true }]
     *   }
     * }, ['id', 'amountInEur', 'description']);
     */
    list(filter, projection) {
        const filterInstance = this.buildPaginatedFilter(filter);
        const projectionInstance = this.buildProjection(projection);
        const rows = this.collection.list(this.caller, filterInstance, projectionInstance);
        return rows;
    }
    /**
     * Update a list of records
     * @param filter the filter that represent the list of records to update
     * @param patch the patch to apply on the selected records
     * @example
     * .update({
     *    conditionTree: {
     *      field: 'isActive',
     *      operator: 'Equal',
     *      value: false
     *    },
     * }, { isActive: true });
     */
    update(filter, patch) {
        const filterInstance = this.buildFilter(filter);
        return this.collection.update(this.caller, filterInstance, patch);
    }
    /**
     * Delete a list of records
     * @param filter the filter that represent the list of records to update
     * @example
     * .delete({
     *    conditionTree: {
     *      field: 'isBlocked',
     *      operator: 'Equal',
     *      value: false,
     *    },
     * });
     */
    delete(filter) {
        const filterInstance = this.buildFilter(filter);
        return this.collection.delete(this.caller, filterInstance);
    }
    /**
     * Aggregate a list of records
     * @param filter the filter used to list the records to aggregate
     * @param aggregation the aggregation to apply
     * @param limit the maximum number of result to return
     * @example
     * .aggregate({
     *    conditionTree: {
     *      field: "user:company:id",
     *      operator: "In",
     *      value: records.map((r) => r.id),
     *    },
     * }, {
     *    operation: "Sum",
     *    field: "amountInEur",
     *    groups: [{ field: "user:company:id" }],
     * }, 10);
     */
    aggregate(filter, aggregation, limit) {
        const filterInstance = this.buildFilter(filter);
        const aggregationInstance = this.buildAggregation(aggregation);
        return this.collection.aggregate(this.caller, filterInstance, aggregationInstance, limit);
    }
    buildFilter(filter) {
        return filter
            ? new unpaginated_1.default({
                ...filter,
                conditionTree: filter.conditionTree
                    ? factory_1.default.fromPlainObject(filter.conditionTree)
                    : undefined,
            })
            : null;
    }
    buildPaginatedFilter(filter) {
        return new paginated_1.default({
            ...filter,
            conditionTree: filter?.conditionTree
                ? factory_1.default.fromPlainObject(filter.conditionTree)
                : undefined,
            sort: filter.sort ? new sort_1.default(...filter.sort) : undefined,
            page: filter.page ? new page_1.default(filter.page.skip, filter.page.limit) : undefined,
        });
    }
    buildProjection(projection) {
        return new projection_1.default(...projection);
    }
    buildAggregation(aggregation) {
        return new aggregation_1.default(aggregation);
    }
}
exports.default = RelaxedCollection;
