// @forestadmin/datasource-mongoose/dist/utils/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.replaceMongoTypes = exports.groupIdsByPath = exports.splitId = exports.compareIds = exports.recursiveSet = exports.escape = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */
const mongoose_1 = require("mongoose");
/**
 * Replace dots by underscores.
 *
 * @example
 * escape('a.b') == 'a_b'
 */
function escape(str) {
    return str.replace(/\./g, '_');
}
exports.escape = escape;
/**
 * Set a value in a plain object recursively
 *
 * @example
 * const a = {};
 * recursiveSet(a, 'field.subfield', 42);
 * a == { field: { subfield: 42 } }
 */
function recursiveSet(target, path, value) {
    const index = path.indexOf('.');
    if (index !== -1) {
        const prefix = path.substring(0, index);
        const suffix = path.substring(index + 1);
        target[prefix] ?? (target[prefix] = {});
        recursiveSet(target[prefix], suffix, value);
    }
    else {
        target[path] = value;
    }
}
exports.recursiveSet = recursiveSet;
/**
 * Compare two ids.
 * This is useful to ensure we perform array operations in the right order.
 *
 * @example
 * compareIds('a.20.a', 'a.1.b') => 1 (because 1 < 20)
 * compareIds('a.0.a', 'b.1.b') => -1 (because 'a' < 'b')
 */
function compareIds(a, b) {
    const isNumber = /^\d+$/;
    const partsA = a.split('.');
    const partsB = b.split('.');
    const length = a.length < b.length ? a.length : b.length;
    for (let i = 0; i < length; i += 1) {
        if (partsA[i] !== partsB[i] && isNumber.test(partsA[i]) && isNumber.test(partsB[i]))
            return Number(partsA[i]) - Number(partsB[i]);
        if (partsA[i] < partsB[i])
            return -1;
        if (partsA[i] > partsB[i])
            return 1;
    }
    if (partsA.length < partsB.length)
        return -1;
    if (partsA.length > partsB.length)
        return 1;
    return 0;
}
exports.compareIds = compareIds;
/**
 * Split the fake ids which are generated when using the `asModel` option into rootId + path.
 *
 * @example
 * splitId('123.some.path.0.to.the.item) == [123, 'some.path.0.to.the.item]
 */
function splitId(id) {
    const dotIndex = id.indexOf('.');
    const rootId = id.substring(0, dotIndex);
    const path = id.substring(dotIndex + 1);
    // @fixme hack, we should never do that without looking at the schema
    return (0, mongoose_1.isValidObjectId)(rootId)
        ? [mongoose_1.Types.ObjectId.createFromHexString(rootId), path]
        : [rootId, path];
}
exports.splitId = splitId;
/**
 * Group the fake ids which are generated when using the `asModel` option by the path they are
 * targeting to help performing requests.
 *
 * @example
 * groupIdsByPath(['123.field', '456.field']) == { field: [123, 456] }
 */
function groupIdsByPath(ids) {
    const updates = {};
    for (const id of ids) {
        const [rootId, path] = splitId(id);
        updates[path] ?? (updates[path] = []);
        updates[path].push(rootId);
    }
    return updates;
}
exports.groupIdsByPath = groupIdsByPath;
/**
 * Replace ObjectId and Date instances in records by strings.
 *
 * @example
 * replaceMongoTypes({_id: ObjectId('12312321313'), createdAt: Date() })
 * == { _id: '12312321313', createdAt: '2010-01-01T00:00:00Z' }
 */
function replaceMongoTypes(records) {
    return JSON.parse(JSON.stringify(records));
}
exports.replaceMongoTypes = replaceMongoTypes;
