// @forestadmin/datasource-toolkit/dist/src/decorators/computed/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const collection_context_1 = __importDefault(require("../../context/collection-context"));
const collection_decorator_1 = __importDefault(require("../collection-decorator"));
const field_1 = __importDefault(require("../../validation/field"));
const compute_fields_1 = __importDefault(require("./helpers/compute-fields"));
const rewrite_projection_1 = __importDefault(require("./helpers/rewrite-projection"));
/** Decorator injects computed fields */
class ComputedCollection extends collection_decorator_1.default {
    constructor() {
        super(...arguments);
        this.computeds = {};
    }
    /** @internal */
    getComputed(path) {
        const index = path.indexOf(':');
        if (index === -1)
            return this.computeds[path];
        const { foreignCollection } = this.schema.fields[path.substring(0, index)];
        const association = this.dataSource.getCollection(foreignCollection);
        return association.getComputed(path.substring(index + 1));
    }
    registerComputed(name, computed) {
        // Check that all dependencies exist and are columns
        for (const field of computed.dependencies) {
            field_1.default.validate(this, field);
        }
        if (computed.dependencies.length <= 0) {
            throw new Error(`Computed field '${this.name}.${name}' must have at least one dependency.`);
        }
        this.computeds[name] = computed;
        this.markSchemaAsDirty();
    }
    async list(caller, filter, projection) {
        const childProjection = projection.replace(path => (0, rewrite_projection_1.default)(this, path));
        const records = await this.childCollection.list(caller, filter, childProjection);
        const context = new collection_context_1.default(this, caller);
        return (0, compute_fields_1.default)(context, this, childProjection, projection, records);
    }
    async aggregate(caller, filter, aggregation, limit) {
        // No computed are used in the aggregation => just delegate to the underlying collection.
        if (!aggregation.projection.some(field => this.getComputed(field))) {
            return this.childCollection.aggregate(caller, filter, aggregation, limit);
        }
        // Fallback to full emulation.
        return aggregation.apply(await this.list(caller, filter, aggregation.projection), caller.timezone, limit);
    }
    refineSchema(childSchema) {
        const schema = { ...childSchema, fields: { ...childSchema.fields } };
        for (const [name, computed] of Object.entries(this.computeds)) {
            schema.fields[name] = {
                columnType: computed.columnType,
                defaultValue: computed.defaultValue,
                enumValues: computed.enumValues,
                filterOperators: new Set(),
                isPrimaryKey: false,
                isReadOnly: true,
                isSortable: false,
                type: 'Column',
            };
        }
        return schema;
    }
}
exports.default = ComputedCollection;
