// @forestadmin/datasource-toolkit/dist/src/decorators/hook/context/update.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HookAfterUpdateContext = exports.InternalHookBeforeUpdateContext = exports.HookBeforeUpdateContext = void 0;
const hook_1 = __importDefault(require("./hook"));
class HookBeforeUpdateContext extends hook_1.default {
    constructor(collection, caller, filter, patch) {
        super(collection, caller);
        this._filter = filter;
        this._patch = patch;
    }
    get filter() {
        return Object.freeze(this._filter);
    }
    get patch() {
        return Object.freeze(this._patch);
    }
}
exports.HookBeforeUpdateContext = HookBeforeUpdateContext;
class InternalHookBeforeUpdateContext extends HookBeforeUpdateContext {
    getFilter() {
        return this._filter;
    }
}
exports.InternalHookBeforeUpdateContext = InternalHookBeforeUpdateContext;
class HookAfterUpdateContext extends HookBeforeUpdateContext {
}
exports.HookAfterUpdateContext = HookAfterUpdateContext;
